### PR TITLE
Backend/feat:dynamic-driver-pool

### PR DESCRIPTION
**Pull request by enigma-commits**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/dev/migrations-read-only/dynamic-offer-driver-app/driver_pool_config.sql [hunk 1]
```diff
@@ -66,4 +66,9 @@ ALTER TABLE atlas_driver_offer_bpp.driver_pool_config ADD COLUMN batch_size_on_r
...
 ------- SQL updates -------
 
-ALTER TABLE atlas_driver_offer_bpp.driver_pool_config ADD COLUMN enable_unified_pooling boolean ;
\ No newline at end of file
+ALTER TABLE atlas_driver_offer_bpp.driver_pool_config ADD COLUMN enable_unified_pooling boolean ;
+
+
+------- SQL updates -------
+
+ALTER TABLE atlas_driver_offer_bpp.driver_pool_config ADD COLUMN dynamic_batch_size integer ;
```
Comment (hkmangla) [inline Backend/dev/migrations-read-only/dynamic-offer-driver-app/driver_pool_config.sql:74]: it should be array type?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs [hunk 2]
```diff
@@ -206,9 +207,9 @@ prepareDriverPoolBatch cityServiceTiers merchant driverPoolCfg searchReq searchT
                   _ -> pure ([], [])
               logDebug $ "SpecialPickupZonePoolBatch goHomeDriversInQueue -" <> show goHomeDriversInQueue
               let goHomeDriversInQueueId = map (\d -> d.driverPoolResult.driverId) goHomeDriversInQueue
-              let normalDriversInQueue = filter (\d -> not (d.driverPoolResult.driverId `elem` goHomeDriversInQueueId)) driversInQueue
+              let normalDriversInQueue = filter (\d -> d.driverPoolResult.driverId `notElem` goHomeDriversInQueueId) driversInQueue
               let normalDriversInQueue' = bookAnyFilters transporterConfig normalDriversInQueue []
-              (_, normalDriversInQueueBatch) <- mkDriverPoolBatch merchantOpCityId_ (take driverPoolCfg.driverBatchSize normalDriversInQueue') intelligentPoolConfig transporterConfig batchSize False
+              (_, normalDriversInQueueBatch) <- mkDriverPoolBatch merchantOpCityId_ (take (driverPoolCfg.dynamicBatchSize V.! max (min ((length driverPoolCfg.dynamicBatchSize) - 1) (batchNum + 1)) 0) normalDriversInQueue') intelligentPoolConfig transporterConfig batchSize False
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:212]: What is this logic, can you make it as a common function?

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/dev/migrations-read-only/dynamic-offer-driver-app/driver_pool_config.sql (hunk 1 above)
Comment (enigma-commits) [reply Backend/dev/migrations-read-only/dynamic-offer-driver-app/driver_pool_config.sql:74]: dsl generated like this

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs (hunk 2 above)
Comment (enigma-commits) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:212]: sure

**Review by hkmangla — APPROVED** [review]

**Pull request merged**